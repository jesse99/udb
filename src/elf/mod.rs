//! Generic ELF file support. These can be both core files and executable files.
//! (The executable files are useful because they contain the bulk of the debugging
//! data needed to interpret the core files.) This module contains generic ELF suport
//! but not the support for stuff like interpreting debug symbols.
//! Quick ELF reference: https://gist.github.com/x0nu11byt3/bcb35c3de461e5fb66173071a2379779
//!
//! ELF files start with an ELF header which includes:
//! * A magic number to identify the file as an ELF file.
//! * The architecture, e.g. Linux AMD x86-64.
//! * The offset to and number of program headers.
//! * The offset to and number of section headers.
//!
//! Program headers identify segments. Segments are used by the OS to load an exe into
//! memory. A program header has type, vaddr, offset, etc. Common types are:
//! * Load - for a core file these are usually memory mapped files for the exe and DLLs, for an exe these are text (CPU instructions) and data (eg statics)
//! * Note - variety of metadata, e.g. process and signal info.
//! * TLS - thread local storage info.
//!
//! Section headers identify sections. Sections are used for static linking and don't
//! appear in core files. Section headers have name, type, vaddr, offset, size, etc.
//! There are a lot of types including for the symbol table, string table, etc.
pub mod elf_file;
pub mod elf_files;
pub mod header;
pub mod io;
pub mod notes;
pub mod sections;
pub mod segments;

pub use elf_file::*;
pub use elf_files::*;
pub use header::*;
pub use io::*;
pub use notes::*;
pub use sections::*;
pub use segments::*;
