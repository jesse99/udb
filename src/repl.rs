//! Definitions for the commands that are used interactively, e.g.
//! `bt` and `info registers`.
use clap::{Args, Parser, Subcommand, ValueEnum};
use std::fmt;

#[derive(Parser)]
#[command(version, about, long_about = None)] // TODO about?
#[command(infer_subcommands(true))] // allow abreviations
pub struct Repl {
    #[command(subcommand)]
    pub command: MainCommand,
}

#[derive(Subcommand)]
pub enum MainCommand {
    /// Show backtrace for the current thread
    Bt,

    /// Show low level information about the core and exe files
    Elf(ElfCommand),

    /// Search memory for a bit pattern
    Find(FindArgs),

    /// Show higher level information about the cored process
    Info(InfoCommand),

    /// Print memory range as hex and ascii
    Hexdump(HexdumpArgs),

    /// Exit udb
    Quit,
}

#[derive(Args)]
pub struct ElfCommand {
    #[clap(subcommand)]
    pub action: ElfAction,
}

#[derive(Args)]
pub struct InfoCommand {
    #[clap(subcommand)]
    pub action: InfoAction,
}

#[derive(Subcommand)]
pub enum ElfAction {
    /// Show ELF header
    Header(ExplainArgs),

    /// Show low level address to line tables
    Line(ElfLineArgs),

    /// Show ELF load segments
    Loads(TableArgs),

    /// Show sections
    Notes(TableArgs),

    /// Show relocations
    Relocations(TableArgs),

    /// Show sections
    Sections(TableArgs),

    /// Show segments
    Segments(TableArgs),

    /// Dump string tables
    Strings(StringsArgs),

    /// Show symbols
    Symbols(TableArgs),
}

#[derive(Subcommand)]
pub enum InfoAction {
    /// Print file and line number for a virtual address
    Line(LineArgs),

    /// Show memory mapped files
    Mapped(TableArgs),

    /// Show information the process associated with the core file
    Process(ExplainArgs),

    /// Show general purpose registers
    Registers(RegistersArgs),

    /// Show information about signals
    Signals(TableArgs),
}

#[derive(Args)]
pub struct ExplainArgs {
    /// Show core info unless there is no core or this is set
    #[arg(long)]
    pub exe: bool,

    /// Explain columns, fields, etc.
    #[arg(short, long)]
    pub explain: bool,
}

// TODO should be able to search for other stuff like ints (need to account for endian)
// TODO provide a way to restrict search area?
#[derive(Args)]
pub struct FindArgs {
    /// Default is to search virtual memory in the core file. When this is enabled all
    /// the bytes in both the exe and the core are searched.
    #[arg(long)]
    pub all: bool,

    /// Search for an UTF-8 string e.g. "the brown fox"
    #[arg(long, group = "filter")]
    pub string: Option<String>,

    /// Optionally hexdump count bytes for each address found
    #[arg(short, long, default_value_t = 0)]
    pub count: usize,

    /// Search for a hex string with spaces ignored, e.g. "ab ac acab"
    #[arg(long, group = "filter")]
    pub hex: Option<String>,

    /// Max number of results to report, 0 for unlimited
    #[arg(short, long, default_value_t = 10, requires = "filter")]
    pub max_results: usize,
}

#[derive(Args)]
pub struct ElfLineArgs {
    /// Number of lines to print in the address => line table.
    #[arg(short, long)]
    #[arg(default_value_t = 20)]
    pub max_lines: usize,
}

#[derive(Args)]
pub struct TableArgs {
    /// Show core info unless there is no core or this is set
    #[arg(long)]
    pub exe: bool,

    /// Explain columns, fields, etc.
    #[arg(short, long)]
    pub explain: bool,

    /// Add column headers
    #[arg(short, long)]
    pub titles: bool,
}

#[derive(Args)]
pub struct RegistersArgs {
    /// Also dump rarely used registers such as segment registers
    #[arg(short, long)]
    pub all: bool,

    /// Show core info unless there is no core or this is set
    #[arg(long)]
    pub exe: bool,

    /// Explain columns, fields, etc.
    #[arg(short, long)]
    pub explain: bool,

    /// Add column headers
    #[arg(short, long)]
    pub titles: bool,
}

#[derive(Args)]
pub struct LineArgs {
    /// A virtual address
    #[arg(value_parser = parse_u64_expr)]
    pub addr: u64,
}

#[derive(Args)]
pub struct HexdumpArgs {
    /// Dump the exe instead of the core file
    #[arg(long)]
    pub exe: bool,

    /// Number of bytes to dump
    #[arg(short, long)]
    #[arg(default_value_t = 16)]
    pub count: usize,

    /// How to display the start of each row
    #[arg(short, long, name = "TYPE")]
    #[arg(default_value_t = HexdumpLabels::None)]
    pub labels: HexdumpLabels,

    /// Treat the value as an offset into the ELF file
    #[arg(long)]
    pub offset: bool,

    /// Defaults to an address
    #[arg(value_parser = parse_u64_expr)]
    pub value: u64,
}

#[derive(Clone, Copy, Eq, PartialEq, ValueEnum)]
pub enum HexdumpLabels {
    /// Show nothing at the start of lines
    None,

    /// Show the address for the first byte on each line
    Addr,

    /// Show the offset from zero for the first byte on each line
    Zero,
}

// TODO add a --limit option to truncate? or just --truncate?
#[derive(Args)]
pub struct StringsArgs {
    /// Section index used to dump just one table
    #[arg(short, long)]
    pub index: Option<usize>,

    /// Max number of results to report for each table, 0 for unlimited
    #[arg(short, long, default_value_t = 10)]
    pub max_results: usize,
}

impl fmt::Display for HexdumpLabels {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        match self {
            HexdumpLabels::None => fmt.write_str("none")?,
            HexdumpLabels::Addr => fmt.write_str("addr")?,
            HexdumpLabels::Zero => fmt.write_str("zero")?,
        }
        Ok(())
    }
}

// TODO this should parse at least simple expressions
fn parse_u64_expr(s: &str) -> Result<u64, String> {
    if s.starts_with("0x") {
        let t = s.trim_start_matches("0x");
        u64::from_str_radix(t, 16).map_err(|_| format!("`{s}` isn't a hex or decimal number"))
    } else {
        s.parse()
            .map_err(|_| format!("`{s}` isn't a hex or decimal number"))
    }
}

// use the open crate to launch off-line docs?
//    maybe a --doc option?
//    would this also be useful for visualization?
//       might be especially good for generated graphical representations
//       maybe memory? or tree data structures, eg pointer refs, or recursive object dumps
