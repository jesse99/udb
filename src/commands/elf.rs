use super::tables::{add_field, add_simple};
use crate::commands::tables::{SimpleTableBuilder, TableBuilder};
use crate::debug::SymbolIndex;
use crate::elf::{
    LoadSegment, MemoryMappedFile, ProgramHeader, SectionHeader, SectionType, StringIndex,
    VirtualAddr,
};
use crate::repl::{DebugArgs, ExplainArgs, StringsArgs};
use crate::utils;
use crate::utils::Styling;
use crate::{elf::ElfFile, elf::ElfFiles, repl::TableArgs};

fn get_file(files: &ElfFiles, exe: bool) -> &ElfFile {
    if exe {
        match &files.exe {
            Some(file) => file,
            None => {
                utils::warn("--exe was used but there is no exe: using core instead");
                files.core.as_ref().unwrap()
            }
        }
    } else {
        match &files.core {
            Some(file) => file,
            None => files.exe.as_ref().unwrap(),
        }
    }
}

pub fn info_debug(files: &ElfFiles, args: &DebugArgs) {
    let file = get_file(files, true);
    match file.get_lines() {
        Some(lines) => {
            for (i, unit) in lines.units.iter().enumerate() {
                println!("compilation unit {i}:");
                println!("   sources:");
                for source in unit.source_files.iter() {
                    match source.length {
                        Some(n) => println!("      {}/{} {} bytes", source.dir, source.file, n),
                        None => println!("      {}/{}", source.dir, source.file),
                    }
                }
                println!("   include paths:");
                for i in unit.include_paths.iter() {
                    println!("      {}", i);
                }
            }
            println!("files:");
            for f in lines.files.iter() {
                println!("   {}", f);
            }
            println!("relative addresses:");
            for (a, v) in lines.lines.iter().take(args.max_lines) {
                let f = &lines.files.get(v.file);
                println!("   0x{:x}  {}:{}:{}", a.start.0, f, v.line, v.column);
            }
            if lines.lines.len() > args.max_lines {
                println!("   ...");
            }
        }
        None => println!("Couldn't find .debug_line section"),
    }
}

pub fn info_header(files: &ElfFiles, args: &ExplainArgs) {
    let mut b = SimpleTableBuilder::new();

    let file = get_file(files, args.exe);
    add_simple!(b, "type", file.header.stype(), "type of ELF file");
    if file.reader.little_endian {
        add_simple!(
            b,
            "little endian",
            file.reader.little_endian,
            "words are being laid out in memory with the most significant byte last"
        );
    } else {
        add_simple!(
            b,
            "little endian",
            file.reader.little_endian,
            "words are being laid out out in memory with the most significant byte first"
        );
    }
    if file.reader.sixty_four_bit {
        add_simple!(
            b,
            "64-bit",
            file.reader.sixty_four_bit,
            "pointers are eight bytes"
        );
    } else {
        add_simple!(
            b,
            "64-bit",
            file.reader.sixty_four_bit,
            "pointers are four bytes"
        );
    }
    add_simple!(
        b,
        "osabi",
        file.header.abi(),
        "the OS the binary was compiled for"
    );
    add_simple!(b, "abiversion", file.header.abiversion, "zero for Linux");
    add_simple!(b, "machine", file.header.machine(), "CPU architecture");
    add_simple!(b, "flags", file.header.flags, "Linux has no defined flags");
    add_simple!(
        b,
        "ph_offset",
        file.header.ph_offset,
        "offset in the ELF file to the Program Header table"
    );
    add_simple!(
        b,
        "num_ph_entries",
        file.header.num_ph_entries,
        "number of entries in the Program Header table"
    );
    add_simple!(
        b,
        "section_offset",
        file.header.section_offset,
        "offset in the ELF file to the section header table"
    );
    add_simple!(
        b,
        "num_section_entries",
        file.header.num_section_entries,
        "number of entries in the section header table"
    );
    add_simple!(
        b,
        "string_table_index",
        file.header.string_table_index,
        "section index containing the string table"
    );
    b.println(args.explain);
}

pub fn info_loads(files: &ElfFiles, args: &TableArgs) {
    pub fn find_file(
        files: &Option<Vec<MemoryMappedFile>>,
        vaddr: VirtualAddr,
    ) -> Option<&MemoryMappedFile> {
        if let Some(maps) = files {
            maps.iter().find(|m| m.vbytes.contains(vaddr))
        } else {
            None
        }
    }

    pub fn is_stack(file: &ElfFile, segment: &LoadSegment) -> bool {
        if let Some(status) = file.find_prstatus() {
            let bottom = status.get_frame_stack_bottom();
            segment.vbytes.contains(bottom)
        } else {
            false
        }
    }

    let mut builder = TableBuilder::new();
    builder.add_col_l("vaddr", "the virtual address the segment starts at");
    builder.add_col_r("vbytes", "the size of the segment in memory");
    builder.add_col_r("flags", "executable, writeable, and/or readable");
    builder.add_col_r(
        "offset",
        "the offset into the ELF file at which the segment appears",
    );
    builder.add_col_r("obytes", "the size of the segment in the core");
    builder.add_col_l(
        "note",
        "path to memory mapped file or how the segment is used",
    );

    let file = get_file(files, args.exe);
    let files = file.get_memory_mapped_files();
    for segment in file.loads.iter() {
        let mut note = String::new();
        if let Some(file) = find_file(files, segment.vbytes.start) {
            note.push_str(&format!("{} ", file.file_name));
        } else if is_stack(file, segment) {
            note.push_str("[stack] ");
        } else if !segment.executable() && segment.writeable() && segment.readable() {
            note.push_str("[data] ");
        } else if !segment.executable() && !segment.writeable() && segment.readable() {
            // TODO may also want to check that the first bytes are '.ELF'
            note.push_str("[text] ");
        }

        add_field!(builder, "vaddr", "{:x}", segment.vbytes.start.0);
        add_field!(builder, "vbytes", "{:x}", segment.vbytes.size);
        add_field!(builder, "flags", segment.flags());
        add_field!(builder, "offset", "{:x}", segment.obytes.start.0);
        add_field!(builder, "obytes", "{:x}", segment.obytes.size);
        add_field!(builder, "note", note);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_notes(files: &ElfFiles, args: &TableArgs) {
    let mut builder = TableBuilder::new();
    builder.add_col_l("name", "note namespace");
    builder.add_col_l("type", "the type of the note");
    builder.add_col_r("offset", "offset into the ELF file (hex)");
    builder.add_col_r("size", "size of the note");

    let file = get_file(files, args.exe);
    for note in file.notes.iter() {
        add_field!(builder, "name", note.name);
        add_field!(builder, "type", "{:?}", note.ntype);
        add_field!(builder, "offset", "{:x}", note.contents.start.0);
        add_field!(builder, "size", note.contents.size);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_relocations(files: &ElfFiles, args: &TableArgs) {
    // TODO probably should use an arg w/o --exe
    let mut builder = TableBuilder::new();
    builder.add_col_r("symbol", "name of the symbol to relocate");
    builder.add_col_r("index", "index of the symbol to relocate"); // TODO get rid of this
    builder.add_col_r("string", "index of the symbol string"); // TODO get rid of this
    builder.add_col_r("dynamic", "true if the symbol is from a shared library");
    builder.add_col_r("offset", "vaddr for exe or shared object");
    builder.add_col_l("type", "how to apply the relocation (arch specific)");
    builder.add_col_r("addend", "optional constant applied during relocation");

    let file = get_file(files, true);
    let symbols = file.find_symbols();
    let dynamic_symbols = file.find_dynamic_symbols();

    let rels = files.find_relocations();
    for r in rels.iter() {
        // TODO names aren't great for static relocation entries. They do match what
        // `readelf --syms` reports but they are sucky names. For example,
        // `readelf --relocs` will report "printf@GLIBC_2.2.5" but we say
        // "deregister_tm_clones".
        let name = if r.dynamic {
            match &dynamic_symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| file.find_string(t.section.link, ue.name))
                }
                None => None,
            }
        } else {
            match &symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| file.find_string(t.section.link, ue.name))
                }
                None => None,
            }
        }
        .flatten()
        .unwrap_or(format!("index {}", r.symbol));

        let string = if r.dynamic {
            match &dynamic_symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| ue.name)
                }
                None => None,
            }
        } else {
            match &symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| ue.name)
                }
                None => None,
            }
        }
        .unwrap_or(StringIndex(0));

        let addend = match r.addend {
            Some(a) => format!("{}", a),
            None => "none".to_string(),
        };
        add_field!(builder, "symbol", name);
        add_field!(builder, "dynamic", r.dynamic);
        add_field!(builder, "index", r.symbol);
        add_field!(builder, "string", string.0);
        add_field!(builder, "offset", "{:x}", r.offset);
        add_field!(builder, "type", "{:?}", r.rtype);
        add_field!(builder, "addend", addend);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_sections(files: &ElfFiles, args: &TableArgs) {
    let file = get_file(files, args.exe);
    let sections = file.get_sections();

    let mut builder = TableBuilder::new();
    builder.add_col_r("index", "index into sections.");
    builder.add_col_l("name", "index into the string table. Zero means no name.");
    builder.add_col_l("type", "type of the section.");
    builder.add_col_r("vaddr", "virtual address at execution.");
    builder.add_col_r(
        "offset",
        "offset into the ELF file for the start of the section.",
    );
    builder.add_col_r("size", "section size in bytes.");
    builder.add_col_r("entry_size", "set if the section holds a table of entries.");
    builder.add_col_r("align", "section alignment.");
    builder.add_col_r(
        "link",
        "link to another section with related information, usually a string or symbol table.",
    );
    builder.add_col_r("info", "additional section info");
    builder.add_col_l("flags", "write, alloc, and/or exec.");

    // Would be kind of nice to sort these by name but they are referenced sometimes
    // by index...
    for (i, section) in sections.iter().enumerate() {
        add_field!(builder, "index", i); // sections are often referenced by index so this is handy
        match file.find_default_string(section.name) {
            Some(n) => {
                add_field!(builder, "name", n);
            }
            None => {
                add_field!(builder, "name", section.name.0);
            }
        };
        add_field!(builder, "type", "{:?}", section.stype);
        add_field!(builder, "flags", SectionHeader::flags(section.flags));
        add_field!(builder, "vaddr", "{:x}", section.vbytes.start.0);
        add_field!(builder, "offset", "{:x}", section.obytes.start.0);
        add_field!(builder, "size", section.vbytes.size);
        add_field!(builder, "entry_size", section.entry_size);
        add_field!(builder, "align", section.align);
        add_field!(builder, "link", section.link.0);
        add_field!(builder, "info", section.info);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_segments(files: &ElfFiles, args: &TableArgs) {
    let file = get_file(files, args.exe);
    let segments = ElfFile::find_segments(file.reader, &file.header);

    let mut builder = TableBuilder::new();
    builder.add_col_l("type", "the segment type");
    builder.add_col_r(
        "offset",
        "the offset into the ELF file at which the segment appears",
    );
    if file.is_core() {
        builder.add_col_r("vaddr", "the virtual address the segment starts at");
    }
    builder.add_col_r("file size", "the size of the segment on disk");
    if file.is_core() {
        builder.add_col_r("memory size", "the size of the segment in memory");
    }
    builder.add_col_r("flags", "executable, writeable, and/or readable");

    for segment in segments.iter() {
        add_field!(builder, "type", "{:?}", segment.stype);
        add_field!(builder, "offset", "{:x}", segment.offset);
        if file.is_core() {
            add_field!(builder, "vaddr", "{:x}", segment.vaddr);
        }
        add_field!(builder, "file size", "{:x}", segment.file_size);
        if file.is_core() {
            add_field!(builder, "memory size", "{:x}", segment.mem_size);
        }
        add_field!(builder, "flags", "{}", ProgramHeader::flags(segment.flags));
    }

    builder.println(args.titles, args.explain);
    if args.explain {
        println!();
        println!("Numeric fields are all in hex. Usually it's more informative to use");
        println!("other commands like `info loads` or `info mapped`.");
    }
}

pub fn info_strings(files: &ElfFiles, args: &StringsArgs) {
    let file = get_file(files, true);
    let num_sections = file.get_sections().len();

    let mut found = false;
    for index in 0..num_sections {
        if args.index.unwrap_or(index) == index {
            let section = &file.get_sections()[index];
            if section.stype == SectionType::StringTable {
                if found {
                    println!();
                }
                println!("section {index}");
                let strings = file.find_strings(section, args.max_results);
                for (i, s) in strings.iter().enumerate() {
                    println!("{i}: {s}");
                }
                if strings.len() == args.max_results {
                    println!("...");
                }
                found = true;
            }
        }
    }
}

pub fn info_symbols(files: &ElfFiles, args: &TableArgs) {
    let mut builder = TableBuilder::new();
    builder.add_col_r("index", "symbol index");
    builder.add_col_l("name", "the symbol name");
    builder.add_col_l("type", "the symbol type");
    builder.add_col_l("dynamic", "true if the symbol is from a shared lib");
    builder.add_col_r("value", "address, absolute value, etc (in hex)");
    builder.add_col_r("size", "size of the value, 0 for unknown or undefined");
    builder.add_col_l("binding", "linkage visibility and behavior");
    builder.add_col_l(
        "visibility",
        "whether the symbol is visible outside its object file",
    );
    builder.add_col_l(
        "related",
        "indicates a related section or marks the entry as an absolute value",
    );

    // TODO double check that function pointers are legit
    // TODO sort rows? provide some sort of generic table sort support?
    // TODO maybe also filtering options, eg max-results and filter by col value
    //      two options for filter by col? or something like --filter="type=Func"?
    //      maybe also a complement option
    let file = get_file(files, args.exe);
    let tables = [file.find_dynamic_symbols(), file.find_symbols()];
    let tables = tables.iter().flatten().collect::<Vec<_>>();
    for table in tables.iter() {
        println!("using section {}", table.section.link.0);
        for (i, e) in table.entries.iter().enumerate() {
            // TODO function names can be really long (especially with name mangling)
            // readelf puts a pretty small cap on these, maybe we should default to the same
            let name = file
                .find_string(table.section.link, e.name)
                .unwrap_or("unknown".to_string());
            let name = format!("{} ({})", name, e.name.0);
            add_field!(builder, "index", i);
            add_field!(builder, "name", name);
            add_field!(builder, "dynamic", table.dynamic);
            add_field!(builder, "value", "{:x}", e.value);
            add_field!(builder, "size", e.size);
            add_field!(builder, "type", "{:?}", e.stype);
            add_field!(builder, "binding", "{:?}", e.binding);
            add_field!(builder, "visibility", "{:?}", e.visibility);
            add_field!(builder, "related", index_to_str(file, e.index));
        }
    }

    builder.println(args.titles, args.explain);
}

fn index_to_str(file: &ElfFile, index: SymbolIndex) -> String {
    match index {
        SymbolIndex::Abs => "Value".to_string(),
        SymbolIndex::Common => "Common".to_string(),
        SymbolIndex::Index(i) => file
            .find_section_name(i)
            .unwrap_or("bad section index".to_string()),
        SymbolIndex::Undef => "".to_string(),
        SymbolIndex::XIndex => "not implemented".to_string(), // TODO
    }
}
