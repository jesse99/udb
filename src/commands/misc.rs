use crate::elf::{ElfFile, ElfOffset, LoadSegment, VirtualAddr};
use crate::repl::HexdumpLabels;
use crate::{
    elf::{ElfFiles, Reader},
    repl::{FindArgs, HexdumpArgs},
    utils,
};
use std::error::Error;

/// Returns pointers to the instructions within the functions in the current call chain.
fn raw_backtrace(files: &ElfFiles) -> Result<Vec<VirtualAddr>, Box<dyn Error>> {
    // TODO move this into debug module
    // see https://eli.thegreenplace.net/2011/09/06/stack-frame-layout-on-x86-64
    let mut bt = Vec::new();
    if let Some(status) = files.find_prstatus() {
        let addr = status.get_ip();
        bt.push(addr);

        let mut rbp = status.get_frame_stack_top(); // TODO won't work for release
        if let Some(load) = files.find_load_segment(rbp)
            && load.writeable()
        {
            // we expect stack to be within one segment
            // TODO could do some validation here but I think we want to be fairly permissive
            while let Some(offset) = load.to_offset(rbp) {
                rbp = VirtualAddr::from_raw(
                    files
                        .core
                        .as_ref()
                        .unwrap() // safe because find_prstatus worked
                        .reader
                        .read_xword(offset.0 as usize)
                        .unwrap(),
                );

                let addr = VirtualAddr::from_raw(
                    files
                        .core
                        .as_ref()
                        .unwrap()
                        .reader
                        .read_xword((offset.0 + 8) as usize)
                        .unwrap(),
                );
                bt.push(addr);
            }
        } else {
            return Err("Couldn't find load segment".into());
        }
    } else {
        return Err("Couldn't find prstatus".into());
    }
    Ok(bt)
}

pub fn backtrace(files: &ElfFiles) {
    match raw_backtrace(files) {
        Ok(bt) => bt.iter().for_each(|a| println!("{:x}", a.0)),
        Err(e) => println!("{e}"),
    }
}

pub fn find(files: &ElfFiles, args: &FindArgs) {
    fn match_bytes(reader: &Reader, i: usize, bytes: &[u8]) -> bool {
        for (j, byte) in bytes.iter().enumerate() {
            match reader.read_byte(i + j) {
                Ok(b) => {
                    if b != *byte {
                        return false;
                    }
                }
                Err(_) => return false,
            }
        }
        true
    }

    fn search_load_segments(core: &ElfFile, args: &FindArgs, bytes: &[u8]) {
        let mut count = 0;
        for load in core.loads.iter() {
            let mut i = 0;
            while i + bytes.len() < load.vbytes.size {
                if match_bytes(&core.reader, i + load.obytes.start.0 as usize, bytes) {
                    println!("0x{:x}", i + load.vbytes.start.0 as usize);
                    if args.count > 0 {
                        hexdump_segment(
                            core,
                            &HexdumpArgs {
                                value: i as u64 + load.vbytes.start.0,
                                offset: false,
                                count: args.count,
                                labels: HexdumpLabels::None,
                                exe: false,
                            },
                            load,
                        );
                        println!();
                    }
                    i += bytes.len();
                    count += 1;
                    if count == args.max_results {
                        println!("...");
                        return;
                    }
                } else {
                    i += 1;
                }
            }
        }
    }

    fn search_all(prefix: &str, file: &ElfFile, args: &FindArgs, bytes: &[u8]) {
        let mut count = 0;
        let mut offset = ElfOffset::from_raw(0);
        let mut offsets = Vec::new(); // we'll print addresses first

        let mut found_addr = false;
        while offset.0 as usize + bytes.len() < file.reader.len() {
            if match_bytes(&file.reader, offset.0 as usize, bytes) {
                match file.find_vaddr(offset) {
                    Some((load, addr)) => {
                        if !found_addr {
                            println!("{prefix}Addresses:");
                            found_addr = true;
                        }
                        println!("   0x{:x}", addr.0);

                        if args.count > 0 {
                            print!("   ");
                            hexdump_segment(
                                file,
                                &HexdumpArgs {
                                    value: addr.0,
                                    offset: false,
                                    exe: false,
                                    count: args.count,
                                    labels: HexdumpLabels::None,
                                },
                                load,
                            );
                            println!();
                        }
                    }
                    None => offsets.push(offset),
                }
                offset = offset + bytes.len() as i64;
                count += 1;
                if count == args.max_results {
                    println!("...");
                    return;
                }
            } else {
                offset = offset + 1;
            }
        }

        if !offsets.is_empty() {
            println!("{prefix}Offsets:");
            for offset in offsets.iter() {
                println!("   0x{:x}", offset.0);

                if args.count > 0 {
                    print!("   ");
                    file.reader
                        .hex_dump(0, offset.0 as usize, args.count, HexdumpLabels::None);
                    println!();
                }
            }
        }
    }

    fn find(files: &ElfFiles, args: &FindArgs, bytes: &[u8]) {
        if args.all {
            if let Some(core) = &files.core
                && let Some(exe) = &files.exe
            {
                search_all("Core ", core, args, bytes);
                search_all("Exe ", exe, args, bytes);
            } else if let Some(core) = &files.core {
                search_all("", core, args, bytes);
            } else {
                search_all("", files.exe.as_ref().unwrap(), args, bytes); // safe because we'll always have either core or exe
            }
        } else if let Some(core) = &files.core {
            search_load_segments(core, args, bytes);
        } else {
            // Technically we should only do this if --all is used but it's kind of
            // silly to not do a search if all we have is an exe.
            search_all("", files.exe.as_ref().unwrap(), args, bytes);
        }
    }

    // TODO there are probably crates with better substring algorithms
    // TODO might also help to read words at a time where possible
    if let Some(s) = &args.hex {
        // TODO need to search both the exe and core (for --all)
        match byte_str_to_vec(s) {
            Ok(bytes) => find(files, args, &bytes),
            Err(err) => utils::warn(&err.to_string()),
        }
    } else if let Some(s) = &args.string {
        let bytes = ascii_str_to_vec(s);
        find(files, args, &bytes);
    }
}

pub fn hexdump(files: &ElfFiles, args: &HexdumpArgs) {
    if args.offset {
        if args.exe {
            match &files.exe {
                Some(file) => hexdump_any(file, args.value as usize, args.count),
                None => utils::warn("--exe was used but there is no exe"),
            }
        } else {
            match &files.core {
                Some(file) => hexdump_any(file, args.value as usize, args.count),
                None => hexdump_any(files.exe.as_ref().unwrap(), args.value as usize, args.count),
            }
        }
    } else {
        let vaddr = VirtualAddr::from_raw(args.value);
        if args.exe {
            match &files.exe {
                Some(file) => match file.find_load_segment(vaddr) {
                    Some(load) => hexdump_segment(file, args, load),
                    None => utils::warn("--couldn't find a load segment for the address"),
                },
                None => utils::warn("--exe was used but there is no exe"),
            }
        } else {
            match &files.core {
                Some(file) => match file.find_load_segment(vaddr) {
                    Some(load) => hexdump_segment(file, args, load),
                    None => utils::warn("couldn't find a load segment for the address"),
                },
                None => {
                    let file = files.exe.as_ref().unwrap();
                    match file.find_load_segment(vaddr) {
                        Some(load) => hexdump_segment(file, args, load),
                        None => utils::warn("couldn't find a load segment for the address"),
                    }
                }
            }
        }
    }
}

pub fn hexdump_segment(file: &ElfFile, args: &HexdumpArgs, load: &LoadSegment) {
    let vaddr = VirtualAddr::from_raw(args.value);
    if let Some(offset) = load.to_offset(vaddr) {
        file.reader
            .hex_dump(args.value, offset.0 as usize, args.count, args.labels);
    }
}

pub fn hexdump_any(file: &ElfFile, offset: usize, count: usize) {
    file.reader.hex_dump(0, offset, count, HexdumpLabels::Zero);
}

fn ascii_str_to_vec(str: &str) -> Vec<u8> {
    let mut result = Vec::new();

    for ch in str.chars() {
        let mut buffer = [0; 4]; // 4 is always large enough
        let n = ch.encode_utf8(&mut buffer).len();
        result.extend(buffer.iter().take(n));
    }

    result
}

fn byte_str_to_vec(str: &str) -> Result<Vec<u8>, Box<dyn Error>> {
    let mut result = Vec::new();

    let mut i = 0;
    let chars: Vec<char> = str.chars().collect();
    while i < chars.len() {
        if chars[i] == ' ' {
            // ignore spaces
            i += 1;
        } else if i + 1 < chars.len()
            && chars[i].is_ascii_hexdigit()
            && chars[i + 1].is_ascii_hexdigit()
        {
            let s = format!("{}{}", chars[i], chars[i + 1]);
            let byte = u8::from_str_radix(&s, 16).unwrap();
            result.push(byte);
            i += 2;
        } else {
            return Err("Expected a string of hex bytes with optional spaces between bytes".into());
        }
    }

    Ok(result)
}
