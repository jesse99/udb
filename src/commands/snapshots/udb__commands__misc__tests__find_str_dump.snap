---
source: src/commands/misc.rs
expression: s
---
0x7ff8fc2ceb25
63 6f 75 6e 74 00 5f 5f  63 6f 75 6e 74 00 5f 5f    count.__rseq_fla
67 73 00 5f 64 6c 5f 61  67                         gs._dl_al

0x7ff8fc3077c4
63 6f 75 6e 74 00 00 00  63 6f 75 6e 74 00 00 00    count...........
00 00 00 00 00 00 00 00  00                         .........

0x7ff8fc3078a9
63 6f 75 6e 74 00 00 00  63 6f 75 6e 74 00 00 00    count...........
00 00 00 00 00 00 00 00  00                         .........


0x7f9c7097eb25
63 6f 75 6e 74 00 5f 5f  63 6f 75 6e 74 00 5f 5f    count.__rseq_fla
67 73 00 5f 64 6c 5f 61  67                         gs._dl_al

0x7f9c709b77c4
63 6f 75 6e 74 00 00 00  63 6f 75 6e 74 00 00 00    count...........
00 00 00 00 00 00 00 00  00                         .........

0x7f9c709b78a9
63 6f 75 6e 74 00 00 00  63 6f 75 6e 74 00 00 00    count...........
00 00 00 00 00 00 00 00  00                         .........
